### PR TITLE
Fix `s/fn` performance issues in Clojure

### PR DESCRIPTION
**Pull request by frenchy64**
While trying to implement `s/defprotocol`, I needed the following invariant to hold in Clojure to access [__methodImplCache](https://github.com/clojure/clojure/blob/b8132f92f3c3862aa6cdd8a72e4e74802a63f673/src/clj/clojure/core_deftype.clj#L609):

```clojure
(let [f (s/fn this [] this)]
  (identical? f (f)))
```

The fix I chose was to use `^{:schema schema} (fn ...)` instead of `(with-meta (fn ...) {:schema schema})`. I believe this also fixed all the documented performance caveats for `s/fn` in Clojure as it means wrappers are eliminated.

The perf characteristics for CLJS seem unchanged (still uses the wrapper). I worked around [CLJS-968](https://clojure.atlassian.net/browse/CLJS-968) by placing the function evaluation in non-tail position. I decided working around this issue was better than using `macros/if-cljs` (which I treat as a last resort).

I also moved a test, please ignore whitespace changes when reviewing.

### CODE REVIEW

File: src/cljc/schema/core.cljc [hunk 1]
```diff
@@ -1276,20 +1276,27 @@
...
+             up to 20 arguments, and then via `apply` beyond 20 arguments.
+             See `cljs.core/with-meta` and `cljs.core.MetaFn`."
   [& fn-args]
   (let [fn-args (if (symbol? (first fn-args))
                   fn-args
                   (cons (gensym "fn") fn-args))
         [name more-fn-args] (macros/extract-arrow-schematized-element &env fn-args)
         {:keys [outer-bindings schema-form fn-body]} (macros/process-fn- &env name more-fn-args)]
-    `(let ~outer-bindings
-       (schematize-fn
```
Comment (w01fe) [inline src/cljc/schema/core.cljc:1290]: Can schematize-fn be deleted/deprecated now?

**Review by w01fe — APPROVED** [review]
Very nice, thanks!

File: src/cljc/schema/core.cljc (hunk 1 above)
Comment (frenchy64) [reply src/cljc/schema/core.cljc:1290]: Looks like it's used by the most recent changes to `s/letfn` :)

Comment (frenchy64) [reply src/cljc/schema/core.cljc:1290]: Hmm, with the knowledge gained here, now the tradeoffs are less obvious about whether to use schematize-fn or declare-class-schema! in `s/letfn`...

Comment (frenchy64) [reply src/cljc/schema/core.cljc:1290]: Ah, perhaps we can have our cake and eat it too by using the same trick in `s/letfn`:

```clojure
(s/letfn [(f [a ...])]
  ...)
=>
(letfn [(f [a ...])]
  (let [f ^{:schema ...} (fn [a ..]))]
    ...))

Comment (frenchy64) [reply src/cljc/schema/core.cljc:1290]: Created an issue to track this https://github.com/plumatic/schema/issues/431

**Comment by frenchy64** [comment]
There was an oversight in `lein all` that meant the latest cljs was never tested. I needed to drop Clojure 1.7 from the test suite because CLJS doesn't work with it (but we only support Clojure 1.8 and later anyway).
